umbraco/Umbraco-CMS
# Implements unattended package migrations for both explicit and implicit migrations

__#10449 Must be reviewed and merged first__

## Changes

* Moves unattended logic to notifications/handlers
* Removes unneeded code like package uninstall and package files
* Adds the actual MigrationPlan instances to the migration context so any step can reference the plan itself (we use this for the implicit migrations)
* Adds package migration expressions to do the actual package install in steps

## Testing

* try the package migrations using code mentioned below
* make one of your steps throw an exception and ensure that the boot failed exception middleware reports the problem

## To create an explicit package migration

An explicit package migration is a package migration created by a package developer that wants full control over the migration process and possibly has custom steps (not just installing the umbraco data from the package.xml file).

```cs
public class TestPackageMigrationPlan : PackageMigrationPlan
{
    public TestPackageMigrationPlan()
        // The name of your package, this will be used
        // to track the packages migration state.
        : base("Articulate") { }

    protected override void DefinePlan()
    {
        // Define migration plan.
        // Each step must have a new unique GUID.
        // Each step maps to a migration class which executes migration logic.
            
        To<TestMigration1>(new Guid("BB02C392-4007-4A6C-A550-28BA2FF7E43D"));

        // The last step listed is the "Final" step.
        // This is the value that will be recorded in the database.
        // If the database value doesn't match this value, the the steps
        // registered in this migration will execute from the last registered
        // step in the database (if any).
        // This is how incremental migrations work (i.e. for upgrades)

        To<TestMigration2>(new Guid("1E3DA142-D5E0-4E3D-A57E-29E8B5B2A9E8"));
    }

    private class TestMigration1 : PackageMigrationBase
    {
        public TestMigration1(IPackagingService packagingService, IMigrationContext context)
            : base(packagingService, context)
        {
        }

        // This is where the logic goes to process a migration step.
        // Logic in here can do all sorts of things like created/modify Database
        // tables or execute c# logic.
        // You can even run logic using the Umbraco services (you need to
        // inject them).
        public override void Migrate()
        {
            // This step will import the Umbraco data (package.xml) file
            // that was exported from the package builder section in the
            // back office.
            // In this case, this is an embedded resource that exists in
            // the same location/namespace as this migration class.
            ImportPackage.FromEmbeddedResource<TestMigration2>().Do();
        }
    }

    private class TestMigration2 : PackageMigrationBase
    {
        public TestMigration2(IPackagingService packagingService, IMigrationContext context) : base(packagingService, context)
        {
        }

        public override void Migrate()
        {
            // The first step already did this, however this is an example
            // of an 'upgrade'. Perhaps the first step previously ran
            // and now you've updated the package.xml manifest with some
            // changes and you want those applied in a new package release.
            // So now you can just execute this step again which will be
            // exactly like you were to upgrade a v8 package in the back office.
            ImportPackage.FromEmbeddedResource<TestMigration2>().Do();
        }
    }
}
```

## To create an implicit package migration

An implicit package migration is for package developers who just want the package.xml installed. To do this, only a single class needs to be created:

```cs
public class TestAutomaticPackageMigrationPlan : AutomaticPackgeMigrationPlan
{
    public TestAutomaticPackageMigrationPlan()
        // The name of your package, this will be used
        // to track the packages migration state.
        : base("MyPackage")
    {
    }
}
```

The requirements for this is that the package.xml file is an embedded resource that exists in the same folder/namespace as this class.
